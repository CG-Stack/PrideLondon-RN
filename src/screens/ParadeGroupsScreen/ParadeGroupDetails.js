// @flow
import React, { PureComponent } from "react";
import type { ElementRef } from "react";
import { Animated, Easing, Linking, StyleSheet, View } from "react-native";
import type { ParadeGroup } from "../../data/parade-group";
import Text from "../../components/Text";
import TextLink from "../../components/TextLink";
import Touchable from "../../components/Touchable";
import text from "../../constants/text";
import chevron from "../../../assets/images/chevronRightGreen.png";

type Props = {
  paradeGroup: ParadeGroup
};

type State = {
  animation: Object,
  collapsed: boolean,
  contentHeight: number
};

const renderURL = (label, url) => (
  <Touchable onPress={() => Linking.openURL(url)}>
    <TextLink>{label}</TextLink>
  </Touchable>
);

const spinInterpolation = {
  inputRange: [0, 1],
  outputRange: ["90deg", "-90deg"]
};

class ParadeGroupDetails extends PureComponent<Props, State> {
  constructor() {
    super();
    // $FlowFixMe: For some reason flow doesn't know about React.createRef.
    this.collapsable = React.createRef();
  }

  state = {
    animation: new Animated.Value(0),
    collapsed: true,
    contentHeight: 0
  };

  componentDidUpdate(prevProps: Props, prevState: State) {
    if (prevState.collapsed === this.state.collapsed) return;

    const toValue = this.state.collapsed ? 0 : 1;
    const { animation } = this.state;

    Animated.timing(animation, {
      toValue,
      duration: 200,
      easing: Easing.ease
    }).start();
  }

  collapsable: ElementRef<typeof View>;

  toggleCollapse = () => {
    if (this.collapsable.current && this.state.collapsed) {
      this.collapsable.current.measure((a, b, width, height) => {
        this.setState({
          collapsed: !this.state.collapsed,
          contentHeight: height
        });
      });
    } else {
      this.setState({
        collapsed: !this.state.collapsed
      });
    }
  };

  render() {
    const {
      name,
      section,
      facebookUrl,
      twitterUrl,
      websiteUrl
    } = this.props.paradeGroup.fields;
    const { animation, collapsed, contentHeight } = this.state;
    const spin = animation.interpolate(spinInterpolation);
    const height = animation.interpolate({
      inputRange: [0, 1],
      outputRange: [4, contentHeight] // must have some height when collapsed to be able to measure correctly.
    });
    return (
      <View style={styles.container}>
        <Touchable
          accessibilityLabel={
            collapsed
              ? text.paradeGroups.expandAccessibilityLabel
              : text.paradeGroups.collapseAccessibilityLabel
          }
          onPress={this.toggleCollapse}
          style={styles.collapseButton}
        >
          <Animated.Image
            style={[styles.chevron, { transform: [{ rotate: spin }] }]}
            source={chevron}
          />
        </Touchable>
        <View style={styles.details}>
          <Text type="h4">{name}</Text>
          <Animated.View style={[styles.animatedContainer, { height }]}>
            <View style={styles.collapsable} ref={this.collapsable}>
              <Text style={styles.section}>{section}</Text>
              {facebookUrl &&
                renderURL(text.paradeGroups.linkLabelFacebook, facebookUrl)}
              {twitterUrl &&
                renderURL(text.paradeGroups.linkLabelTwitter, twitterUrl)}
              {websiteUrl &&
                renderURL(text.paradeGroups.linkLabelWebsite, websiteUrl)}
            </View>
          </Animated.View>
        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 0,
    flexGrow: 1,
    flexShrink: 1,
    flexDirection: "row",
    alignItems: "flex-start"
  },
  chevron: {
    width: 24,
    height: 24
  },
  collapseButton: {},
  details: {
    flex: 0,
    flexGrow: 1,
    flexShrink: 1,
    flexDirection: "column",
    paddingTop: 10,
    paddingBottom: 10
  },
  animatedContainer: {
    overflow: "hidden"
  },
  collapsable: {
    minHeight: 32 // required to stop text from collapsing
  },
  section: {
    paddingTop: 8
  }
});

export default ParadeGroupDetails;
