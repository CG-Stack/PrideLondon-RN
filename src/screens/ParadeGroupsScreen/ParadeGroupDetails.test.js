// @flow
import React from "react";
import { shallow } from "enzyme";
import Touchable from "../../components/Touchable";
import ParadeGroupDetails from "./ParadeGroupDetails";

describe("when collapsed", () => {
  it("renders correctly with all fields", () => {
    const paradeGroup = {
      contentType: "paradeGroup",
      id: "ABC",
      locale: "en-GB",
      revision: 1,
      fields: {
        name: "A Test Parade Group",
        section: "Section A",
        facebookUrl: "https://red-badger.com",
        twitterUrl: "https://red-badger.com",
        websiteUrl: "https://red-badger.com"
      }
    };
    const output = shallow(<ParadeGroupDetails paradeGroup={paradeGroup} />);
    expect(output).toMatchSnapshot();
  });

  it("renders correctly with missing fields", () => {
    const paradeGroup = {
      contentType: "paradeGroup",
      id: "DEF",
      locale: "en-GB",
      revision: 1,
      fields: {
        name: "Test Parade Group",
        section: "Section B",
        facebookUrl: null,
        twitterUrl: null,
        websiteUrl: null
      }
    };
    const output = shallow(<ParadeGroupDetails paradeGroup={paradeGroup} />);
    expect(output).toMatchSnapshot();
  });
});

describe("when expanded", () => {
  it("renders correctly", () => {
    const paradeGroup = {
      contentType: "paradeGroup",
      id: "ABC",
      locale: "en-GB",
      revision: 1,
      fields: {
        name: "A Test Parade Group",
        section: "Section A",
        facebookUrl: "https://red-badger.com",
        twitterUrl: "https://red-badger.com",
        websiteUrl: "https://red-badger.com"
      }
    };
    const output = shallow(<ParadeGroupDetails paradeGroup={paradeGroup} />);
    const mockRef = {
      measure: jest.fn()
    };
    output.instance().collapsable.current = mockRef;
    output
      .find(Touchable)
      .first()
      .simulate("press");
    mockRef.measure.mock.calls[0][0](0, 0, 0, 100);
    expect(output).toMatchSnapshot();
    expect(output.instance().state.collapsed).toEqual(false);
    expect(output.instance().state.contentHeight).toEqual(100);
  });
});

describe("when expanded then collapsed", () => {
  it("renders correctly", () => {
    const paradeGroup = {
      contentType: "paradeGroup",
      id: "ABC",
      locale: "en-GB",
      revision: 1,
      fields: {
        name: "A Test Parade Group",
        section: "Section A",
        facebookUrl: "https://red-badger.com",
        twitterUrl: "https://red-badger.com",
        websiteUrl: "https://red-badger.com"
      }
    };
    const output = shallow(<ParadeGroupDetails paradeGroup={paradeGroup} />);
    const mockRef = {
      measure: jest.fn()
    };
    output.instance().collapsable.current = mockRef;
    output
      .find(Touchable)
      .first()
      .simulate("press");
    mockRef.measure.mock.calls[0][0](0, 0, 0, 100);
    output
      .find(Touchable)
      .first()
      .simulate("press");
    expect(output).toMatchSnapshot();
    expect(output.instance().state.collapsed).toEqual(true);
  });
});
