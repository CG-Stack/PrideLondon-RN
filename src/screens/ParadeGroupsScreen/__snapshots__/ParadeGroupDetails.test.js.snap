// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when collapsed renders correctly with all fields 1`] = `
<Component
  style={
    Object {
      "alignItems": "flex-start",
      "flex": 0,
      "flexDirection": "row",
      "flexGrow": 1,
      "flexShrink": 1,
    }
  }
>
  <Touchable
    accessibilityComponentType="button"
    accessibilityLabel="Show more"
    accessibilityTraits={
      Array [
        "button",
      ]
    }
    delayPressIn={50}
    onPress={[Function]}
    style={Object {}}
  >
    <AnimatedComponent
      source={1}
      style={
        Array [
          Object {
            "height": 24,
            "width": 24,
          },
          Object {
            "transform": Array [
              Object {
                "rotate": "90deg",
              },
            ],
          },
        ]
      }
    />
  </Touchable>
  <Component
    style={
      Object {
        "flex": 0,
        "flexDirection": "column",
        "flexGrow": 1,
        "flexShrink": 1,
        "paddingBottom": 10,
        "paddingTop": 10,
      }
    }
  >
    <Text
      color="blackColor"
      markdown={false}
      markdownStyle={Object {}}
      type="h4"
    >
      A Test Parade Group
    </Text>
    <AnimatedComponent
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "height": 4,
          },
        ]
      }
    >
      <Component
        style={
          Object {
            "minHeight": 32,
          }
        }
      >
        <Text
          color="blackColor"
          markdown={false}
          markdownStyle={Object {}}
          style={
            Object {
              "paddingTop": 8,
            }
          }
          type="text"
        >
          Section A
        </Text>
        <Touchable
          accessibilityComponentType="button"
          accessibilityTraits={
            Array [
              "button",
            ]
          }
          delayPressIn={50}
          onPress={[Function]}
        >
          <TextLink>
            Facebook
          </TextLink>
        </Touchable>
        <Touchable
          accessibilityComponentType="button"
          accessibilityTraits={
            Array [
              "button",
            ]
          }
          delayPressIn={50}
          onPress={[Function]}
        >
          <TextLink>
            Twitter
          </TextLink>
        </Touchable>
        <Touchable
          accessibilityComponentType="button"
          accessibilityTraits={
            Array [
              "button",
            ]
          }
          delayPressIn={50}
          onPress={[Function]}
        >
          <TextLink>
            Website
          </TextLink>
        </Touchable>
      </Component>
    </AnimatedComponent>
  </Component>
</Component>
`;

exports[`when collapsed renders correctly with missing fields 1`] = `
<Component
  style={
    Object {
      "alignItems": "flex-start",
      "flex": 0,
      "flexDirection": "row",
      "flexGrow": 1,
      "flexShrink": 1,
    }
  }
>
  <Touchable
    accessibilityComponentType="button"
    accessibilityLabel="Show more"
    accessibilityTraits={
      Array [
        "button",
      ]
    }
    delayPressIn={50}
    onPress={[Function]}
    style={Object {}}
  >
    <AnimatedComponent
      source={1}
      style={
        Array [
          Object {
            "height": 24,
            "width": 24,
          },
          Object {
            "transform": Array [
              Object {
                "rotate": "90deg",
              },
            ],
          },
        ]
      }
    />
  </Touchable>
  <Component
    style={
      Object {
        "flex": 0,
        "flexDirection": "column",
        "flexGrow": 1,
        "flexShrink": 1,
        "paddingBottom": 10,
        "paddingTop": 10,
      }
    }
  >
    <Text
      color="blackColor"
      markdown={false}
      markdownStyle={Object {}}
      type="h4"
    >
      Test Parade Group
    </Text>
    <AnimatedComponent
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "height": 4,
          },
        ]
      }
    >
      <Component
        style={
          Object {
            "minHeight": 32,
          }
        }
      >
        <Text
          color="blackColor"
          markdown={false}
          markdownStyle={Object {}}
          style={
            Object {
              "paddingTop": 8,
            }
          }
          type="text"
        >
          Section B
        </Text>
      </Component>
    </AnimatedComponent>
  </Component>
</Component>
`;

exports[`when expanded renders correctly 1`] = `
<Component
  style={
    Object {
      "alignItems": "flex-start",
      "flex": 0,
      "flexDirection": "row",
      "flexGrow": 1,
      "flexShrink": 1,
    }
  }
>
  <Touchable
    accessibilityComponentType="button"
    accessibilityLabel="Show more"
    accessibilityTraits={
      Array [
        "button",
      ]
    }
    delayPressIn={50}
    onPress={[Function]}
    style={Object {}}
  >
    <AnimatedComponent
      source={1}
      style={
        Array [
          Object {
            "height": 24,
            "width": 24,
          },
          Object {
            "transform": Array [
              Object {
                "rotate": "90deg",
              },
            ],
          },
        ]
      }
    />
  </Touchable>
  <Component
    style={
      Object {
        "flex": 0,
        "flexDirection": "column",
        "flexGrow": 1,
        "flexShrink": 1,
        "paddingBottom": 10,
        "paddingTop": 10,
      }
    }
  >
    <Text
      color="blackColor"
      markdown={false}
      markdownStyle={Object {}}
      type="h4"
    >
      A Test Parade Group
    </Text>
    <AnimatedComponent
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "height": 4,
          },
        ]
      }
    >
      <Component
        style={
          Object {
            "minHeight": 32,
          }
        }
      >
        <Text
          color="blackColor"
          markdown={false}
          markdownStyle={Object {}}
          style={
            Object {
              "paddingTop": 8,
            }
          }
          type="text"
        >
          Section A
        </Text>
        <Touchable
          accessibilityComponentType="button"
          accessibilityTraits={
            Array [
              "button",
            ]
          }
          delayPressIn={50}
          onPress={[Function]}
        >
          <TextLink>
            Facebook
          </TextLink>
        </Touchable>
        <Touchable
          accessibilityComponentType="button"
          accessibilityTraits={
            Array [
              "button",
            ]
          }
          delayPressIn={50}
          onPress={[Function]}
        >
          <TextLink>
            Twitter
          </TextLink>
        </Touchable>
        <Touchable
          accessibilityComponentType="button"
          accessibilityTraits={
            Array [
              "button",
            ]
          }
          delayPressIn={50}
          onPress={[Function]}
        >
          <TextLink>
            Website
          </TextLink>
        </Touchable>
      </Component>
    </AnimatedComponent>
  </Component>
</Component>
`;

exports[`when expanded then collapsed renders correctly 1`] = `
<Component
  style={
    Object {
      "alignItems": "flex-start",
      "flex": 0,
      "flexDirection": "row",
      "flexGrow": 1,
      "flexShrink": 1,
    }
  }
>
  <Touchable
    accessibilityComponentType="button"
    accessibilityLabel="Show more"
    accessibilityTraits={
      Array [
        "button",
      ]
    }
    delayPressIn={50}
    onPress={[Function]}
    style={Object {}}
  >
    <AnimatedComponent
      source={1}
      style={
        Array [
          Object {
            "height": 24,
            "width": 24,
          },
          Object {
            "transform": Array [
              Object {
                "rotate": "90deg",
              },
            ],
          },
        ]
      }
    />
  </Touchable>
  <Component
    style={
      Object {
        "flex": 0,
        "flexDirection": "column",
        "flexGrow": 1,
        "flexShrink": 1,
        "paddingBottom": 10,
        "paddingTop": 10,
      }
    }
  >
    <Text
      color="blackColor"
      markdown={false}
      markdownStyle={Object {}}
      type="h4"
    >
      A Test Parade Group
    </Text>
    <AnimatedComponent
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "height": 4,
          },
        ]
      }
    >
      <Component
        style={
          Object {
            "minHeight": 32,
          }
        }
      >
        <Text
          color="blackColor"
          markdown={false}
          markdownStyle={Object {}}
          style={
            Object {
              "paddingTop": 8,
            }
          }
          type="text"
        >
          Section A
        </Text>
        <Touchable
          accessibilityComponentType="button"
          accessibilityTraits={
            Array [
              "button",
            ]
          }
          delayPressIn={50}
          onPress={[Function]}
        >
          <TextLink>
            Facebook
          </TextLink>
        </Touchable>
        <Touchable
          accessibilityComponentType="button"
          accessibilityTraits={
            Array [
              "button",
            ]
          }
          delayPressIn={50}
          onPress={[Function]}
        >
          <TextLink>
            Twitter
          </TextLink>
        </Touchable>
        <Touchable
          accessibilityComponentType="button"
          accessibilityTraits={
            Array [
              "button",
            ]
          }
          delayPressIn={50}
          onPress={[Function]}
        >
          <TextLink>
            Website
          </TextLink>
        </Touchable>
      </Component>
    </AnimatedComponent>
  </Component>
</Component>
`;
